tests/test_database_matrix.py

```diff
@@ -0,0 +1,571 @@
+        ],
+    )
+
+
+@pytest.mark.asyncio
+async def test_update_same_key_single_state_key(patched_send, opsdroid_matrix):
+    patched_send.return_value = nio.RoomGetStateEventResponse(
+        {"twim": {"hello": "world"}}, "", "", ""
+    )
+
+    value = {"hello": "bob"}
+    db = DatabaseMatrix({"single_state_key": True}, opsdroid=opsdroid_matrix)
+    db.should_migrate = False
+
+    await db.put("twim", value)
+
+    patched_send.assert_has_calls(
+        [
+            matrix_call(
+                "GET",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/",
+            ),
+            matrix_call(
+                "PUT",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/",
+                {"twim": value},
+            ),
+        ],
+    )
+
+
+@pytest.mark.asyncio
+async def test_default_update_same_key_value(patched_send, opsdroid_matrix):
+    patched_send.return_value = nio.RoomGetStateEventResponse(
+        {"hello": "world"}, "", "", ""
+    )
+
+    value = {"hello": "world"}
+    db = DatabaseMatrix({"single_state_key": False}, opsdroid=opsdroid_matrix)
+    db.should_migrate = False
+
+    await db.put("twim", value)
+
+    patched_send.assert_has_calls(
+        [
+            matrix_call(
+                "GET",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/twim",
+            )
+        ],
+    )
+
+
+@pytest.mark.asyncio
+async def test_default_update_same_key_value_single_state_key(
+    patched_send, opsdroid_matrix
+):
+    patched_send.return_value = nio.RoomGetStateEventResponse(
+        {"twim": {"hello": "world"}}, "", "", ""
+    )
+
+    value = {"hello": "world"}
+    db = DatabaseMatrix({"single_state_key": True}, opsdroid=opsdroid_matrix)
+    db.should_migrate = False
+
+    await db.put("twim", value)
+
+    patched_send.assert_has_calls(
+        [
+            matrix_call(
+                "GET",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/",
+            )
+        ],
+    )
+
+
+@pytest.mark.asyncio
+async def test_default_update_single_state_key(patched_send, opsdroid_matrix):
+    patched_send.return_value = nio.RoomGetStateEventResponse(
+        {"twim": "hello"}, "", "", ""
+    )
+
+    db = DatabaseMatrix({"single_state_key": True}, opsdroid=opsdroid_matrix)
+    db.should_migrate = False
+
+    await db.put("pill", "red")
+
+    patched_send.assert_has_calls(
+        [
+            matrix_call(
+                "GET",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/",
+            ),
+            matrix_call(
+                "PUT",
+                "/_matrix/client/r0/rooms/%21notaroomid/state/dev.opsdroid.database/",
+                {"twim": "hello", "pill": "red"},
+            ),
+        ],
+    )
+
+
+@pytest.mark.asyncio
+async def test_put_encrypted(patched_send, opsdroid_matrix, mocker):
```

Comment:
I think we need a single_state_key variant of this test, with more than one item in the dict. (probably good to have a couple of tests with more than one item in the dict).